lib/_http_outgoing.js

```diff
@@ -511,11 +532,24 @@ OutgoingMessage.prototype.end = function(data, encoding, callback) {
     throw new TypeError('first argument must be a string or Buffer');
   }
 
+  var self = this;
+
   if (this.finished) {
+    if (data && data.length > 0) {
+      //If the user tried to write data, tell him it failed
+      var err = new Error('write after end');
+      process.nextTick(function() {
+        self.emit('error', err);
+        if (callback) callback(err);
+      });
+    } else {
+      //If he only wanted to end the stream anyways, lucky for him
+      if (callback)
```

Comment:
Above here, you check for `typeof callback === 'function'` rather than a truthy check. Could we stick to one?

Strictly speaking, I don't think there's a perf difference between them.